### PR TITLE
Problem: token mapping state is not exported to genesis

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #101

- add fields in genesis and init/export genesis logic

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [x] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [x] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [x] Have you rebased your work on top of the latest master? 
- [x] Have you checked your code compiles? (`make`)
- [x] Have you included tests for any non-trivial functionality?
- [x] Have you checked your code passes the unit tests? (`make test`)
- [x] Have you checked your code formatting is correct? (`go fmt`)
- [x] Have you checked your basic code style is fine? (`golangci-lint run`)
- [x] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [x] If your changes affect the client infrastructure, have you run the integration test?
- [x] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [x] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [x] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: x/cronos/genesis.go [hunk 1]
```diff
@@ -4,12 +4,22 @@ import (
...
+	"github.com/ethereum/go-ethereum/common"
 )
 
 // InitGenesis initializes the capability module's state from a provided genesis
 // state.
 func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
 	k.SetParams(ctx, genState.Params)
+
+	for _, m := range genState.ExternalContracts {
+		k.SetExternalContractForDenom(ctx, m.Denom, common.HexToAddress(m.Contract))
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/cronos/genesis.go:24]: sounds good, done.

[the remark this answers is not in the document]
Comment (thomas-nguy) [reply x/cronos/genesis.go:24]: can you also complete the unit tests for the added validation logic

[the remark this answers is not in the document]
Comment (yihuang) [reply x/cronos/genesis.go:24]: done

**Review by thomas-nguy — APPROVED** [review]
lgtm

**Comment by Martknka** [comment]
NO

út 28. 9. 2021 v 3:58 odesílatel yihuang ***@***.***> napsal:

**Pull request merged**